src/RequestImpl.cpp

```diff
@@ -292,10 +291,92 @@ class RequestImpl : public Request {
+    template <typename protocolT>
+    boost::asio::ip::tcp::endpoint ToEp(const std::string& endp,
+                                        const protocolT& protocol,
+                                        Context& ctx) const {
+        string host, port;
+
+        auto endipv6 = endp.find(']');
+        if (endipv6 == string::npos) {
+            endipv6 = 0;
+        }
+        const auto pos = endp.find(':', endipv6);
+        if (pos == string::npos) {
+            host = endp;
+        } else {
+            host = endp.substr(0, pos);
+            if (endp.size() > pos) {
+                port = endp.substr(pos + 1);
+            }
+        }
+
+        // Strip IPv6 brackets, asio can't resolve it
+        if (endipv6 && !host.empty() && host.front() == '[') {
+            host = host.substr(1, endipv6 -1);
+        }
+
+        RESTC_CPP_LOG_TRACE_("host=" << host << ", port=" << port);
+
+        if (host.empty()) {
+            const auto port_num = stoi(port);
+            if (port_num < 1 || port_num > std::numeric_limits<uint16_t>::max()) {
+                throw RestcCppException{"Port number out of range: "s + port};
+            }
+            return {protocol, static_cast<uint16_t>(port_num)};
+        }
+
+        boost::asio::ip::tcp::resolver::query q{host, port};
+        boost::asio::ip::tcp::resolver resolver(owner_.GetIoService());
+
+        auto ep = resolver.async_resolve(q, ctx.GetYield());
+        const decltype(ep) addr_end;
+        for(; ep != addr_end; ++ep)
+        if (ep != addr_end) {
+
+            RESTC_CPP_LOG_TRACE_("ep=" << ep->endpoint() << ", protocol=" << ep->endpoint().protocol().protocol());
+
+            if (protocol == ep->endpoint().protocol()) {
+                return ep->endpoint();
+            }
+
+            RESTC_CPP_LOG_TRACE_("Incorrect protocol, looping for next alternative");
+        }
+
+        RESTC_CPP_LOG_ERROR_("Failed to resolve endpoint " << endp
+                             << " with protocol " << protocol.protocol());
+        throw FailedToResolveEndpointException{"Failed to resolve endpoint: "s + endp};
+    }
+
+
+    auto GetBindProtocols(const std::string& endp, Context& ctx) {
+        std::vector<decltype(boost::asio::ip::tcp::v4())> p;
+
+        if (!endp.empty()) {
+            try {
+                ToEp(endp, boost::asio::ip::tcp::v4(), ctx);
+                p.push_back(boost::asio::ip::tcp::v4());
+            } catch (const FailedToResolveEndpointException&) {
+                ;
+            }
+
+            try {
+                ToEp(endp, boost::asio::ip::tcp::v6(), ctx);
+                p.push_back(boost::asio::ip::tcp::v6());
+            } catch (const FailedToResolveEndpointException&) {
+                ;
+            }
+        }
+
+        return p;
+    }
+
     Connection::ptr_t Connect(Context& ctx) {
 
         static const auto timer_name = "Connect"s;
 
+        auto prot_filter = GetBindProtocols(properties_->bindToLocalAddress, ctx);
```

Comment:
Only call `GetBindProtocols()`  under `if (!connection->GetSocket().IsOpen())` to save cost?

Reply:
It's the work-flow for new connections. All connection object instances are maintained by a connection-pool, and this block opens the connection if it is new.

Reply:
Yes, but `Connect()` is called for every `SendRequest()`, so `GetBindProtocols()` is called for every time also. So I suggest to only call `GetBindProtocols()` if `!connection->GetSocket().IsOpen()`, i.e. move line 378 to line 411

Reply:
I thought we could move this line to line 411 to save cost if the connection is already open, but there is a loop in line 396.